runtime/src/message_processor.rs

```diff
@@ -5,75 +5,58 @@ use serde::{Deserialize, Serialize};
 use solana_sdk::{
     account::{create_keyed_readonly_accounts, Account, KeyedAccount},
     clock::Epoch,
+    entrypoint_native::InvokeContext,
     instruction::{CompiledInstruction, InstructionError},
     message::Message,
     native_loader,
     pubkey::Pubkey,
+    rent::Rent,
     system_program,
     transaction::TransactionError,
 };
 use std::{cell::RefCell, rc::Rc};
 
 // The relevant state of an account before an Instruction executes, used
 // to verify account integrity after the Instruction completes
-#[derive(Debug)]
+#[derive(Clone, Debug, Default)]
 pub struct PreAccount {
-    pub is_writable: bool,
-    pub lamports: u64,
-    pub data_len: usize,
-    pub data: Option<Vec<u8>>,
-    pub owner: Pubkey,
-    pub is_executable: bool,
-    pub rent_epoch: Epoch,
+    key: Pubkey,
+    is_signer: bool,
+    is_writable: bool,
+    is_executable: bool,
+    lamports: u64,
+    data_len: usize,
+    data: Vec<u8>,
+    owner: Pubkey,
+    rent_epoch: Epoch,
 }
 impl PreAccount {
-    pub fn new(account: &Account, program_id: &Pubkey, is_writable: bool) -> Self {
+    pub fn new(key: &Pubkey, account: &Account, is_signer: bool, is_writable: bool) -> Self {
```

Comment:
That puts the burden on the caller to clone it if they intend to use it after the call.  Are you trying to save a possible clone if the caller doesn't intend to use it after the call, I can see that as more efficient

Reply:
Yes, a while back, Rob changed every `Pubkey` parameter in the codebase to `&Pubkey`. Instead, we should have defined the rule that if the code is going to store the value in a struct, you should pass it by value, such that the compiler will use move semantics instead of doing a copy.  If we removed the `Copy` from the `Pubkey` (and `Signature`) definitions, it'd make it a lot more obvious to do this. Any time you passed by reference, you'd see an explicit `clone()`.

Reply:
Got it, thanks for the context! Sounds like this isn't an issue worth addressing in this PR then